### PR TITLE
Docker inspect command exits with error if input objects are of mixed…

### LINKED ISSUE #26493: RFC: docker inspect shouldn't mix types
https://github.com/docker/docker/pull/23614 added support for inspecting all type of objects through the top-level `docker inspect` subcommand. However, when specifying multiple "names" or "ID's", the output can include multiple type of objects, for example:

```bash
docker run -v vol:/somewhere --name cont -dit busybox

docker inspect vol cont busybox
WARNING: --size ignored for volume
```

(not sure where the `--size` warning comes from, given that no `--size` is provided)

Results in:

```
[
    {
        "Name": "vol",
        "Driver": "local",
        "Mountpoint": "/var/lib/docker/volumes/vol/_data",
        "Labels": null,
        "Scope": "local"
    },
    {
        "Id": "36d8b327f701db6cc84b55e68debcf27c748da5f3ecb7b48abc3bf0e641dde5c",
        "Created": "2016-09-12T15:34:48.792009328Z",
        "Path": "sh",
        "Args": [],
        "State": {
            "Status": "running",
            "Running": true,
            "Paused": false,
            "Restarting": false,
            "OOMKilled": false,
            "Dead": false,
            "Pid": 638,
            "ExitCode": 0,
            "Error": "",
            "StartedAt": "2016-09-12T15:34:49.636880124Z",
            "FinishedAt": "0001-01-01T00:00:00Z"
        },
        "Image": "sha256:2b8fd9751c4c0f5dd266fcae00707e67a2545ef34f9a29354585f93dac906749",
        "ResolvConfPath": "/var/lib/docker/containers/36d8b327f701db6cc84b55e68debcf27c748da5f3ecb7b48abc3bf0e641dde5c/resolv.conf",
        "HostnamePath": "/var/lib/docker/containers/36d8b327f701db6cc84b55e68debcf27c748da5f3ecb7b48abc3bf0e641dde5c/hostname",
        "HostsPath": "/var/lib/docker/containers/36d8b327f701db6cc84b55e68debcf27c748da5f3ecb7b48abc3bf0e641dde5c/hosts",
        "LogPath": "/var/lib/docker/containers/36d8b327f701db6cc84b55e68debcf27c748da5f3ecb7b48abc3bf0e641dde5c/36d8b327f701db6cc84b55e68debcf27c748da5f3ecb7b48abc3bf0e641dde5c-json.log",
        "Name": "/cont",
        "RestartCount": 0,
        "Driver": "aufs",
        "MountLabel": "",
        "ProcessLabel": "",
        "AppArmorProfile": "",
        "ExecIDs": null,
        "HostConfig": {
            "Binds": [
                "vol:/somewhere"
            ],
            "ContainerIDFile": "",
            "LogConfig": {
                "Type": "json-file",
                "Config": {}
            },
            "NetworkMode": "default",
            "PortBindings": {},
            "RestartPolicy": {
                "Name": "no",
                "MaximumRetryCount": 0
            },
            "AutoRemove": false,
            "VolumeDriver": "",
            "VolumesFrom": null,
            "CapAdd": null,
            "CapDrop": null,
            "Dns": [],
            "DnsOptions": [],
            "DnsSearch": [],
            "ExtraHosts": null,
            "GroupAdd": null,
            "IpcMode": "",
            "Cgroup": "",
            "Links": null,
            "OomScoreAdj": 0,
            "PidMode": "",
            "Privileged": false,
            "PublishAllPorts": false,
            "ReadonlyRootfs": false,
            "SecurityOpt": null,
            "UTSMode": "",
            "UsernsMode": "",
            "ShmSize": 67108864,
            "Runtime": "runc",
            "ConsoleSize": [
                0,
                0
            ],
            "Isolation": "",
            "CpuShares": 0,
            "Memory": 0,
            "CgroupParent": "",
            "BlkioWeight": 0,
            "BlkioWeightDevice": null,
            "BlkioDeviceReadBps": null,
            "BlkioDeviceWriteBps": null,
            "BlkioDeviceReadIOps": null,
            "BlkioDeviceWriteIOps": null,
            "CpuPeriod": 0,
            "CpuQuota": 0,
            "CpusetCpus": "",
            "CpusetMems": "",
            "Devices": [],
            "DiskQuota": 0,
            "KernelMemory": 0,
            "MemoryReservation": 0,
            "MemorySwap": 0,
            "MemorySwappiness": -1,
            "OomKillDisable": false,
            "PidsLimit": 0,
            "Ulimits": null,
            "CpuCount": 0,
            "CpuPercent": 0,
            "IOMaximumIOps": 0,
            "IOMaximumBandwidth": 0
        },
        "GraphDriver": {
            "Name": "aufs",
            "Data": null
        },
        "Mounts": [
            {
                "Name": "vol",
                "Source": "/var/lib/docker/volumes/vol/_data",
                "Destination": "/somewhere",
                "Driver": "local",
                "Mode": "z",
                "RW": true,
                "Propagation": "rprivate"
            }
        ],
        "Config": {
            "Hostname": "36d8b327f701",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "Tty": true,
            "OpenStdin": true,
            "StdinOnce": false,
            "Env": [
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
            ],
            "Cmd": [
                "sh"
            ],
            "Image": "busybox",
            "Volumes": null,
            "WorkingDir": "",
            "Entrypoint": null,
            "OnBuild": null,
            "Labels": {}
        },
        "NetworkSettings": {
            "Bridge": "",
            "SandboxID": "d09d8dd26616b60d9c51be2dc96940fe079b78b19d00d7ab21456a9050db541e",
            "HairpinMode": false,
            "LinkLocalIPv6Address": "",
            "LinkLocalIPv6PrefixLen": 0,
            "Ports": {},
            "SandboxKey": "/var/run/docker/netns/d09d8dd26616",
            "SecondaryIPAddresses": null,
            "SecondaryIPv6Addresses": null,
            "EndpointID": "0af532d46d5e12b592fac784e4addbdd189253567ecb3f7252818cde5b8bfdb8",
            "Gateway": "172.18.0.1",
            "GlobalIPv6Address": "",
            "GlobalIPv6PrefixLen": 0,
            "IPAddress": "172.18.0.2",
            "IPPrefixLen": 16,
            "IPv6Gateway": "",
            "MacAddress": "02:42:ac:12:00:02",
            "Networks": {
                "bridge": {
                    "IPAMConfig": null,
                    "Links": null,
                    "Aliases": null,
                    "NetworkID": "a183ad260bd219d496e577bf36628b2974cc88ed051fc30fcab2c7cfff76c5c6",
                    "EndpointID": "0af532d46d5e12b592fac784e4addbdd189253567ecb3f7252818cde5b8bfdb8",
                    "Gateway": "172.18.0.1",
                    "IPAddress": "172.18.0.2",
                    "IPPrefixLen": 16,
                    "IPv6Gateway": "",
                    "GlobalIPv6Address": "",
                    "GlobalIPv6PrefixLen": 0,
                    "MacAddress": "02:42:ac:12:00:02"
                }
            }
        }
    },
    {
        "Id": "sha256:2b8fd9751c4c0f5dd266fcae00707e67a2545ef34f9a29354585f93dac906749",
        "RepoTags": [
            "busybox:latest"
        ],
        "RepoDigests": [
            "busybox@sha256:a59906e33509d14c036c8678d687bd4eec81ed7c4b8ce907b888c607f6a1e0e6"
        ],
        "Parent": "",
        "Comment": "",
        "Created": "2016-06-23T23:23:37.198943461Z",
        "Container": "764ef4448baa9a1ce19e4ae95f8cdd4eda7a1186c512773e56dc634dff208a59",
        "ContainerConfig": {
            "Hostname": "55cd1f8f6e5b",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
            ],
            "Cmd": [
                "/bin/sh",
                "-c",
                "#(nop) CMD [\"sh\"]"
            ],
            "Image": "sha256:e732471cb81a564575aad46b9510161c5945deaf18e9be3db344333d72f0b4b2",
            "Volumes": null,
            "WorkingDir": "",
            "Entrypoint": null,
            "OnBuild": null,
            "Labels": {}
        },
        "DockerVersion": "1.10.3",
        "Author": "",
        "Config": {
            "Hostname": "55cd1f8f6e5b",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
            ],
            "Cmd": [
                "sh"
            ],
            "Image": "sha256:e732471cb81a564575aad46b9510161c5945deaf18e9be3db344333d72f0b4b2",
            "Volumes": null,
            "WorkingDir": "",
            "Entrypoint": null,
            "OnBuild": null,
            "Labels": {}
        },
        "Architecture": "amd64",
        "Os": "linux",
        "Size": 1092588,
        "VirtualSize": 1092588,
        "GraphDriver": {
            "Name": "aufs",
            "Data": null
        },
        "RootFS": {
            "Type": "layers",
            "Layers": [
                "sha256:8ac8bfaff55af948c796026ee867448c5b5b5d9dd3549f4006d9759b25d4a893"
            ]
        }
    }
]
```

The output contains JSON for a *volume*, an *image*, and a *container*. I don't think that's very useful (especially without a standard property that indicates what type of object an entry is).

Possible improvements;

- If multiple ID's / names are specified, only return objects of the same type as the first ID / name.
- Have a "type" property in every type of object returned (e.g.`type: volume`)

/cc @icecrime @dnephin wdyt?

### PR DESCRIPTION
**Pull request by adshmh**
Add a check to docker inspect command to avoid generating output for objects of different types. Fixes #26493

Signed-off-by: Arash Deshmeh <adeshmeh@ca.ibm.com>

**- What I did**
The docker inspect command will output JSON for different types of objects (e.g. container, network, etc.) if the objects specified as its input are of different types. This PR fixes this by generating an error if the types of objects specified as input to the command are not all the same.

**- How I did it**
Added a check to the inspect command, to exit with an error on the first object name that is not of the same type as the previous one(s) specified as the command's input. This is done before generating output on any object's attributes.

**- How to verify it**
Create 2 or more objects of different types,specify all their names as the inspect command's input. 
With this PR, the command exits with an error, e.g.:

% docker network create net1
% docker volume create vol1

% docker inspect net1 vol1
Error: all objects to inspect were expected to be of type network (same as net1), but vol1 is of type volume

**- Description for the changelog**
The docker inspect command exits with error if the input names correspond to objects of different types. 

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by adshmh** [comment]
The reason for proposing the change to the integration test TestInspectStopWhenNotFound() is that with this PR, the command will generate the error message for "not-shown" reference and exit before generating any output for the "busybox" reference.

**Comment by adshmh** [comment]
Updated the PR, changed the test to create a Volume instead of a Network, as the test was failing on Windows (it seems this can happen on Windows: #23314)

**Comment by LK4D4** [comment]
I think everyone agreed on design in #26493

**Comment by LK4D4** [comment]
ping @icecrime @cpuguy83 @vdemeester

**Comment by LK4D4** [comment]
ping @docker/core-engine-maintainers

File: cli/command/system/inspect.go
Comment (thaJeztah) [inline cli/command/system/inspect.go]: wondering if we could find a "cleaner" solution; the current implementation basically filters/inspects "outside" of the `inspect.Inspect()` method, and passes the results in.

@vdemeester do you have suggestions?

Another thing I noticed, is that detection if Swarm is enabled is performed for _each_ reference (not something introduced by this PR). Perhaps that's something we could optimize, and perform that check _once_ and keep the results (not for this PR)

Comment (vdemeester) [inline cli/command/system/inspect.go]: Hum yeah, both on this PR and the old code, `/info` is called for each reference, that could be optimized.

> wondering if we could find a "cleaner" solution; the current implementation basically filters/inspects "outside" of the inspect.Inspect() method, and passes the results in.

It's not that big of a deal, it's almost already the case (the definition of the inspect is outside of `inspect` package), so we can do pretty much anything we want 😛

Comment (adshmh) [inline cli/command/system/inspect.go]: @thaJeztah, @vdemeester if needed, I can try moving the code that performs the type check, inside inspect.Inspect(). For that, perhaps inspect.GetRefFunc needs to be enhanced to include the returned object's type and existing calls to Inspect() need to be modified to supply that information.

I can submit a separate PR for optimizing the code that detects if Swarm is enabled.

**Comment by adshmh** [comment]
ping @crosbymichael

**Comment by adshmh** [comment]
@thaJeztah , @vdemeester would appreciate it if you could take another look to see if I should make changes to this PR. 
#30648 seems to mention this PR (i.e. different object types), but perhaps the two PRs are not contradictory (one intends to show all the errors, while this one makes it an error to mix object types)

**Comment by thaJeztah** [comment]
@adshmh sorry for the delay; I tested this PR, but it looks like we're not there yet;

Create some mixed objects (volume, network);

```bash
$ docker volume create foobar
foobar
$docker volume create something
something
$ docker network create something
abe83ec7251959c1d8ab5aef7baa7fd766891b4d9ce422c6134082ff376cfb76
```

Inspect both `foobar` and `something`;

```bash
$ docker inspect foobar something
Error: all objects to inspect were expected to be of type volume (same as foobar), but something is of type network
```

However..

```bash
$ docker inspect something
[
    {
        "Name": "something",
        "Id": "abe83ec7251959c1d8ab5aef7baa7fd766891b4d9ce422c6134082ff376cfb76",
        "Created": "2017-03-06T12:26:25.345633998Z",
        "Scope": "local",
        "Driver": "bridge",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": {},
            "Config": [
                {
                    "Subnet": "172.19.0.0/16",
                    "Gateway": "172.19.0.1"
                }
            ]
        },
        "Internal": false,
        "Attachable": false,
        "Containers": {},
        "Options": {},
        "Labels": {}
    }
]
```

That last inspect should return an error, because there's both a _volume_ and a _network_ named `something`

**Comment by adshmh** [comment]
Thank you for reviewing the PR. If I understand the test cases correctly, the PR needs to be modified to produce errors on names that refer to objects of different types (I will update the PR to do this). 

Regarding errors, should the error on ambiguity of an object name take precedence over the error of mixing object types? i.e. in the following command:
$ docker inspect foobar something

should the error be something like 'ambiguous object name: "something", is a name for multiple objects of types: network, volume'?

**Comment by thaJeztah** [comment]
Hm, I think we can simply say that the inspect returned multiple types, so that the user has to specify a `--type` to only inspect the type that's needed

**Comment by adshmh** [comment]
Thank you for the clarification. I have updated the PR (one additional change I needed to make was to avoid fail fast, to be consistent with PR #30648.

Here is how the inspect command behaves with the updated PR:
$ docker volume create foobar
foobar
$ docker volume create something
something
$ docker network create something
cb9b0633d8a2e964ac83c60bfca920e11dda916fa82fe1154ed59be6674793c1
$ docker network create n1
a0955e8e0bb942996403fc553a2d191582eaa02c88bf0d0d964154f0326847c6

-------------------------------- multiple types for one name:
$ docker inspect foobar something
[
    {
        "Driver": "local",
        "Labels": {},
        "Mountpoint": "/var/lib/docker/volumes/foobar/_data",
        "Name": "foobar",
        "Options": {},
        "Scope": "local"
    }
]
Error: multiple types returned for object "something", use --type option to inspect only the type that is needed.
-------------------------------- multiple types for one name with --type option
$ docker inspect --type volume foobar something
[
    {
        "Driver": "local",
        "Labels": {},
        "Mountpoint": "/var/lib/docker/volumes/foobar/_data",
        "Name": "foobar",
        "Options": {},
        "Scope": "local"
    },
    {
        "Driver": "local",
        "Labels": {},
        "Mountpoint": "/var/lib/docker/volumes/something/_data",
        "Name": "something",
        "Options": {},
        "Scope": "local"
    }
]
------------------------------------------------ mixed object types:
$ docker inspect foobar n1       
[
    {
        "Driver": "local",
        "Labels": {},
        "Mountpoint": "/var/lib/docker/volumes/foobar/_data",
        "Name": "foobar",
        "Options": {},
        "Scope": "local"
    }
]
Error: all objects to inspect were expected to be of type volume (same as foobar), but n1 is of type network

**Comment by adshmh** [comment]
The PR only has an integration test, but no unit tests as there are currently no test files for cli/command/system/inspect.go. But considering that there seem to be a few scenarios, perhaps once the overall logic looks OK I need to add a few unit tests as well.

**Comment by adshmh** [comment]
rebased as I was getting some CI errors which did not seem related to the PR itself.

**Comment by vdemeester** [comment]
@adshmh I think it's related, because now inspect fail if there is two element named the same, and it didn't before. I think we should update the failing test to make sure we use `--type` *or* that there is no 2 different object named the same for this test 👼 

```
04:12:46 ----------------------------------------------------------------------
04:12:46 FAIL: docker_cli_network_unix_test.go:571: DockerNetworkSuite.TestDockerNetworkConnectDisconnect
04:12:46 
04:12:46 docker_cli_network_unix_test.go:581:
04:12:46     c.Assert(waitRun("test"), check.IsNil)
04:12:46 ... value *errors.fundamental = error executing docker inspect: Error: multiple types returned for object "test", use --type option to inspect only the type that is needed.
04:12:46 
04:12:46 
04:12:46  ("error executing docker inspect: Error: multiple types returned for object \"test\", use --type option to inspect only the type that is needed.\n\n\n")
04:12:46 
04:14:33 
04:14:33 ----------------------------------------------------------------------
```

**Comment by adshmh** [comment]
PR is updated, it now includes the error when there are multiple types for an object name, as well as corrections to integration tests (needed as @vdemeester pointed out, inspect now exits with error on  object name with multiple types). The Windows build failure does not seem to be due to changes to this PR  (I got the same error on Windows build in an unrelated PR).

**Comment by vdemeester** [comment]
@adshmh needs another rebase, sorry 😅

**Review by vdemeester — APPROVED** [review]
LGTM 👼 
We can refactor the code (for the swarm detection) in a follow-up PR 👼

**Comment by adshmh** [comment]
@vdemeester thank you for the review. I will follow up with a PR for the swarm mode once this PR is merged.

**Comment by thaJeztah** [comment]
Wondering if this is correct;

```
docker inspect foobar something
[
    {
        "Driver": "local",
        "Labels": {},
        "Mountpoint": "/var/lib/docker/volumes/foobar/_data",
        "Name": "foobar",
        "Options": {},
        "Scope": "local"
    }
]
Error: multiple types returned for object "something", use --type option to inspect only the type that is needed.
```

As it still prints the first result; I think it should probably only print the error 😅

**Comment by adshmh** [comment]
@thaJeztah , thank you for testing the PR. I tried to make the behavior consistent with #30648 (i.e. output for all 'valid' object references and report errors at the end of the output). Please let me know if for this PR no output should be generated if there are any errors.

**Comment by cpuguy83** [comment]
Looks like this needs to move to the docker/cli repo.
Closing since this doesn't apply to this repo anymore.

Let me know if you need any help here.
Thanks!

**Pull request closed without merging**
